### PR TITLE
MDEV-20945: BACKUP UNLOCK + FTWRL assertion failure | SIGSEGV in I_P_List from MDL_context::release_lock on INSERT w/ BACKUP LOCK (on optimized builds) | Assertion `ticket->m_duration == MDL_EXPLICIT' failed.

### PR DESCRIPTION
**Pull request by i-rinat**
During its operation, `THD::leave_locked_tables_mode()` changes all locks to have `MDL_TRANSACTION` duration. Some locks that need to stay `MDL_EXPLICIT` are then on one-by-on basis are changed to have an explicit duration. Since the lock from a `BACKUP LOCK` should only be removed by a corresponding `BACKUP UNLOCK`, this patch changes the duration of the lock too.

### CODE REVIEW

File: mysql-test/main/backup_locks.result [hunk 1]
```diff
@@ -39,6 +39,27 @@ MDL_INTENTION_EXCLUSIVE	Schema metadata lock	test
...
+# Scenario 2.
+CREATE TABLE t1(a INT);
+CREATE TABLE t2(b INT);
+LOCK TABLES t2 AS a2 WRITE;
+BACKUP LOCK t1;
+UNLOCK TABLES;
+INSERT INTO t1 VALUES(0);
+BACKUP UNLOCK;
+DROP TABLE t1;
+DROP TABLE t2;
```
Comment (an3l) [inline mysql-test/main/backup_locks.result:61]: In jira there are more other examples of usage on non-existing tables/databases, would be good to cover them also.

File: sql/sql_class.cc [hunk 2]
```diff
@@ -5604,6 +5604,8 @@ void THD::leave_locked_tables_mode()
       mysql_ha_set_explicit_lock_duration(this);
     if (ull_hash.records)
       mysql_ull_set_explicit_lock_duration(this);
+    if (mdl_backup_lock)
+      mdl_context.set_lock_duration(mdl_backup_lock, MDL_EXPLICIT);
```
Comment (svoj) [inline sql/sql_class.cc:5608]: There's broader problem, please see a comment around https://github.com/MariaDB/server/blob/10.4/sql/lock.cc#L998

In other words FTWRL is expecting connection to hold no meta-data locks, otherwise it may fall into a deadlock which cannot be ruled out even by the mdl deadlock detector. An example of such deadlock is described in https://jira.mariadb.org/browse/MDEV-20946.

The only good fix I can think of is to allow `BACKUP LOCK` only under `BACKUP STAGE.

Comment (krizhanovsky) [reply sql/sql_class.cc:5608]: @svoj thank you for the comment! Do you mean that the fix introduces the new deadlock scenarios or are you about generic solution for MDEV-20946, not this one? We didn't find any new deadlock scenarios introduced with the fix.

We love the proposal
>  to allow `BACKUP LOCK` only under `BACKUP STAGE`
to fix MDEV-20946 - the lock is intended only for backups (I think). But it seems this change will affect the user API. We reviewed https://mariadb.com/kb/en/backup-lock/ and https://mariadb.com/kb/en/backup-stage/ - it seems the change won't break anything which was described as allowed by the KBs, but the locking scenarios from MDEV-20945 and MDEV-20946 won't work. Is such change allowed for 10.5 and earlier releases?

@vuvova could you please comment on this? Are we good to go with the change not only for 10.6, but for current 10.5 and stable 10.4?

Comment (svoj) [reply sql/sql_class.cc:5608]: > Do you mean that the fix introduces the new deadlock scenarios 
@krizhanovsky, no, I mean the change introduced by this PR will likely become a dead-code when the broader problem is fixed. There're no new deadlocks introduced by the fix for sure.

By the "behaviour change" you mean not being able to issue `BACKUP LOCK` out of `BACKUP STAGE`? IMHO it should be fine.

I'm unsure if @vuvova has followed all this backup effort, it'd be nice if @montywi could comment on this.

File: mysql-test/main/backup_locks.result (hunk 1 above)
Comment (i-rinat) [reply mysql-test/main/backup_locks.result:83]: Thanks for the suggestion. Added all snippets I found in Jira's MDEV-20495 as test cases.
Also while checking scenarios I realized that some of the crashes were happening during shutdown, so added server restarts to the test too.

File: sql/sql_class.cc (hunk 2 above)
Comment (vuvova) [reply sql/sql_class.cc:5608]: Yes, @svoj is right, I didn't follow it that much, better ask @montywi 

But https://mariadb.com/kb/en/backup-lock/ does say, though
> One should use different connections for BACKUP STAGE commands and BACKUP LOCK. 

which implies that you cannot only allow `BACKUP LOCK` under `BACKUP STAGE`

Comment (svoj) [reply sql/sql_class.cc:5608]: Gosh, I missed that. We need to figure out the reasoning, I can't see any immediately. @vlad-lesin is `mariabackup` making use of `BACKUP LOCK`. If yes, are those called in a separate connection? At which `BACKUP STAGE`?

Comment (i-rinat) [reply sql/sql_class.cc:5608]: Assuming `BACKUP STAGE` family is used, is there a need to use `BACKUP LOCK`? There is already `BACKUP STAGE BLOCK_DDL` that should block DDL. `BACKUP LOCK` looks redundant in such scenario.

`BACKUP LOCK` itself looks very strange. https://mariadb.com/kb/en/backup-lock/ assumes that data copying should be performed after `BACKUP UNLOCK`. Wouldn't copying the data without any locks cause a potential data corruption? There could be writes to the data store, since nothing prevents that.

Comment (svoj) [reply sql/sql_class.cc:5608]: @i-rinat, IIRC InnoDB tables are being copied during `BACKUP STAGE START` when DDL is permitted. `BACKUP LOCK` was supposed to keep .frm and InnoDB data file in sync.

I don't completely understand the need for this statement either given that we have DDL logger.

Comment (i-rinat) [reply sql/sql_class.cc:5608]: > IIRC InnoDB tables are being copied during BACKUP STAGE START when DDL is permitted. BACKUP LOCK was supposed to keep .frm and InnoDB data file in sync.

So it's just another way to backup InnoDB tables? Besides `BACKUP STAGE`? Latter seems like a completed feature set, while `BACKUP LOCK` is a tool for a narrow specific task of backuping InnoDB tables?

Sorry, I just can't wrap my head around it. If somebody uses `BACKUP STAGE`, why would they use BACKUP LOCK at all? [kb/en/mariabackup-and-backup-stage-commands](https://mariadb.com/kb/en/mariabackup-and-backup-stage-commands/#mariabackup-and-backup-stage-commands-in-mariadb-enterprise-server) describes backup steps. They look self-sufficient, with no additional locking needed. So `BACKUP LOCK` would be of no use in `BACKUP STAGE` family. If it is to be restricted to be called only inside `BACKUP STAGE`'s, then `BACKUP LOCK` may as well be removed completely.

**Comment by i-rinat** [comment]
I believe that crashes listed in MDEV-20945 were caused by wrong lock duration, which is fixed in this pull requests. Since possible deadlocks are not part of MDEV-20945, I'll try to address them in another pull request.

**Comment by i-rinat** [comment]
Made a #1572 for [MDEV-20946](https://jira.mariadb.org/browse/MDEV-20946) (Hard FTWRL deadlock under user level locks). However I currently believe that test timeouts observed in MDEV-20946 were caused by the way `mysqltest` executes instructions in tests and the way the test was written. See the issue and the PR for details.

File: sql/sql_class.cc (hunk 2 above)
Comment (montywi) [reply sql/sql_class.cc:5608]: The usage of backup lock is, as said above, to keep .frm and table files in sync.
The ddl log is not usable to detect this case as we use to .frm to check with the ddl log if
the table was copied before or after the ddl.  If all table files are not in sync, then mariabackup could think it got the latest version of the tables, which it didn't

Comment (montywi) [reply sql/sql_class.cc:5608]: It's correct that we do unlock before copying the tables. This works as we open the table BEFORE doing an unlock. If there is any alter table while we are copying the tables, it will not affect the backup as we are working with the original tables

Comment (montywi) [reply sql/sql_class.cc:5608]: I have updated MDEV-20945 how to do a proper fix.
Can you please do a new patch based on the given suggestions?

Comment (i-rinat) [reply sql/sql_class.cc:5608]: > do a new patch based on the given suggestions?

I'll make the changes.

**Comment by i-rinat** [comment]
Made BACKUP LOCK/UNLOCK statement to cause an implicit commit and prevented BACKUP LOCK/UNLOCK from being used in stored procedures. As for prohibiting DDL, it was easier to reuse what's already in the code to prevent any data modifications from the connection that issues BACKUP LOCK. It's highly unlikely anybody will start to insert data after acquiring BACKUP LOCK.

File: sql/sql_class.cc (hunk 2 above)
Comment (i-rinat) [reply sql/sql_class.cc:5608]: > do a new patch based on the given suggestions

Updated the patch.
It was a bit easier to also forbid data modifications too. Should be fine though, as it's unlikely that anybody would want to insert data while under BACKUP LOCK.

**Comment by i-rinat** [comment]
@montywi, ping?

**Comment by i-rinat** [comment]
> give an error for using FLUSH under BACKUP LOCK.

Updated the patch as were suggested: FLTWRL/LOCK TABLES are now mutually exclusive with BACKUP LOCK.

> The change in leave_locked_tables_mode() looks wrong.

It looks wrong, but it wasn't wrong. `THD::leave_locked_tables_mode()` calls `mdl_context.set_transaction_duration_for_all_locks()` which forcefully changes `mdl_backup_lock` duration to `MDL_TRANSACTION`. It does that to all other locks too. Then some locks are updated to be `MDL_EXPLICIT` since they were explicit before and needs to stay explicit. Omitting `mdl_backup_lock` there breaks symmetry. Other locks that should have their duration `MDL_EXPLICIT`, are updated, but `mdl_backup_lock` is not.

I don't have a strong preference about adding it, since mutual exclusion of locked tables mode and BACKUP LOCK should prevent that code from running anyway. So, deleted it.

Probably, the most clean solution will be do something about `set_explicit_duration_for_all_locks()` which is called in `THD::enter_locked_tables_mode()`. For example, introduce an "original" duration for locks so that they could be reset to that state.

**Comment by mariadb-RoelVandePaar** [comment]
@i-rinat did you check all testcases in MDEV-20945 / MDEV-22448 against your patch? Do they work afterwards? Thank you!

**Comment by i-rinat** [comment]
@mariadb-RoelVandePaar,

> did you check all testcases in MDEV-20945 / MDEV-22448 against your patch? Do they work afterwards?

Patch contains testcases derived from examples from MDEV-20945 (up to the comment from up to 2020-05-27 05:56). Most of them repeat themselves. They repeat the same sequence of events: issue BACKUP LOCK, execute something that enters locked tables mode, then leaves locked tables mode (`THD::leave_locked_tables_mode()`, where lock from BACKUP LOCK is scheduled for too-early release), and then issue BACKUP UNLOCK. Assertion triggers in `backup_unlock()`, which called by either issuing BACKUP UNLOCK directly, or BACKUP LOCK that implicitly unlocks before trying to lock, or by shutting down the server process, which implicitly unlocks acquired locks.

I've also just checked every case from comments of MDEV-20945 after the comment from 2020-05-27 05:56. I couldn't reproduce crashes or assertion failures with them. Should I add them as testcases too? As mentioned before, they are essentially the same as those already added.

> Also, any idea on all the failures? https://buildbot.mariadb.org/#/builders/25/builds/7114

There is a report for that issue: https://jira.mariadb.org/browse/MDBF-34.
Unfortunately, I have no idea what could be the root cause for that. I never was able to reproduce anything similar locally.
First few times I encountered that SIGPIPE issue, I tried to look closely at all logs from a buildbot run, but found no clues. Everything looks like something suddenly decided to send a SIGPIPE to mysql-test-run.pl.

**Comment by mariadb-RoelVandePaar** [comment]
@i-rinat I think there may be some merit in adding all testcases as I usually only include other testcases when they generates at least a partially different stack. Still, if per-testcase you know that it is a clear repeat, feel free to not do so offcourse. I will test all of them before removing the filter after this is approved + merged. Thank you for your efforts on this!

**Comment by i-rinat** [comment]
> there may be some merit in adding all testcases as I usually only include other testcases when they generates at least a partially different stack.

It's hard to say anything specific until this PR is merged, but I still think most of the testcases are the same, as they follow similar paths. Locking, spurious lock release, actual lock release with a failed assertion. Release versions suffer from memory corruption, which may cause crashes in other places, resulting in different stack backtraces.

It's worth to add that each testcase which involve a server restart, adds another 1-2 seconds to tests run time. Once added, tests are rarely removed, so adding another tests that are most probably duplicates, will slow down whole tests a bit without adding anything valuable.

**Comment by mariadb-RoelVandePaar** [comment]
@i-rinat understood, thank you!

File: sql/sql_reload.cc [hunk 3]
```diff
@@ -216,6 +216,7 @@ bool reload_acl_and_cache(THD *thd, unsigned long long options,
               !thd->mdl_context.has_locks() ||
               thd->handler_tables_hash.records ||
               thd->ull_hash.records ||
+              thd->mdl_backup_lock ||
```
Comment (montywi) [inline sql/sql_reload.cc:219]: Move the test of mdl_backup_lock one line down to be close to thd->current_backup_stage as these both test are related to backup

File: sql/sql_reload.cc [hunk 4]
```diff
@@ -516,6 +517,14 @@ bool flush_tables_with_read_lock(THD *thd, TABLE_LIST *all_tables)
   Lock_tables_prelocking_strategy lock_tables_prelocking_strategy;
   TABLE_LIST *table_list;
 
+  /*
```
Comment (montywi) [inline sql/sql_reload.cc:520]: Move this a few lines down, before:
  if (thd->current_backup_stage != BACKUP_FINISHED)
  {
    my_error(ER_BACKUP_LOCK_IS_ACTIVE, MYF(0));
    return true;
  }

Same reason as previous comment:  Code is easier to understand if all backup related tests are close to each other

**Review by montywi — CHANGES_REQUESTED** [review]
Please fix the few things that I had in my comments and I will then push it into MariaDB
If you don't want/cannot  do them, I can do them for you before pushing.

File: sql/backup.cc [hunk 5]
```diff
@@ -379,6 +379,10 @@ bool backup_reset_alter_copy_lock(THD *thd)
 bool backup_lock(THD *thd, TABLE_LIST *table)
 {
   backup_unlock(thd);
+  if (thd->locked_tables_mode) {
```
[the remark this answers is not in the document]
Comment (i-rinat) [reply sql/backup.cc]: Looks like I had `{` this way in multiple places in the patch. Updated them all.

File: sql/sql_class.h [hunk 6]
```diff
@@ -3626,7 +3626,8 @@ class THD: public THD_count, /* this must be first */
   inline bool has_read_only_protection()
   {
     if (current_backup_stage == BACKUP_FINISHED &&
-        !global_read_lock.is_acquired())
```
[the remark this answers is not in the document]
Comment (i-rinat) [reply sql/sql_class.h:3629]: Added a related testcase to `main.backup_locks`.

File: sql/sql_reload.cc (hunk 3 above)
Comment (i-rinat) [reply sql/sql_reload.cc]: Moved as suggested.

File: sql/sql_reload.cc (hunk 4 above)
Comment (i-rinat) [reply sql/sql_reload.cc]: Moved this one too, as suggested.

**Comment by i-rinat** [comment]
Updated the patch according to the suggestions from the review.

And while changes were made, decided to update `--error` directives to use readable error names rather than numbers. I was asked to do similar change in other pull requests, and guessed that this is the preferred way. If that was a wrong decision, please notify me, and I'll change that part back.

File: sql/backup.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (montywi) [reply sql/backup.cc]: thanks

**Review by montywi — APPROVED** [review]
Looks good. Will push into 10.4 shortly!

**Comment by an3l** [comment]
Pushed with 709ba7dcae871e365697f to 10.4. Thanks
